### PR TITLE
adding libcurl to support pause/resume download

### PR DESCRIPTION
**Pull request by namchuai**
## Describe Your Changes

- Refactor, remove post processing download (extract, parse yaml, etc.) from download service.
- Adding percentage download (default from `libcurl`)

## Remaining things to do
- [ ] Haven't tested with `cortex update`

## Fixes Issues

- Closes #
- Closes #

## Self Checklist

- [ ] Added relevant comments, esp in complex areas
- [ ] Updated docs (for bug fixes / features)
- [ ] Created issues for follow-up changes or refactoring needed

### CODE REVIEW

File: engine/utils/url_parser.h [hunk 1]
```diff
@@ -0,0 +1,84 @@
+#include <regex>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <vector>
+#include "exceptions/malformed_url_exception.h"
+
+namespace url_parser {
```
Comment (namchuai) [inline engine/utils/url_parser.h:9]: self note: write unit test

File: engine/commands/cortex_upd_cmd.h [hunk 2]
```diff
@@ -11,7 +10,7 @@ namespace commands {
 #ifndef CORTEX_VARIANT
 #define CORTEX_VARIANT file_manager_utils::kProdVariant
 #endif
-constexpr const auto kNightlyHost = "https://delta.jan.ai";
+constexpr const auto kNightlyHost = "delta.jan.ai";
```
Comment (dan-menlo) [inline engine/commands/cortex_upd_cmd.h:13]: Should these be pulled out into a `.cortexrc`? It seems very fragile to hard code a lot of these? cc @vansangpfiev @nguyenhoangthuan99

File: engine/utils/cortexso_parser.h [hunk 3]
```diff
@@ -6,6 +6,7 @@
 #include <services/download_service.h>
 #include <nlohmann/json.hpp>
 #include "httplib.h"
+#include "utils/file_manager_utils.h"
 
 namespace cortexso_parser {
```
Comment (dan-menlo) [inline engine/utils/cortexso_parser.h:11]: What are we trying to refer to by naming this `cortexso_parser`? 
We might need a better name for this - i.e. this refers to a Cortex Model Repo (i.e. Git repo, tag-based)

**Review by dan-menlo — COMMENTED** [review]
@namchuai I will let @vansangpfiev and @nguyenhoangthuan99 review this, but left a couple of comments on naming and architecture.

**Review by vansangpfiev — APPROVED** [review]

File: engine/utils/cortexso_parser.h (hunk 3 above)
Comment (namchuai) [reply engine/utils/cortexso_parser.h:11]: I agree. But I will KIV this. Ideally, we will have multiple url parsers.

File: engine/commands/cortex_upd_cmd.h (hunk 2 above)
Comment (namchuai) [reply engine/commands/cortex_upd_cmd.h:13]: `.cortexrc` should be for things that are configurable for user? I doubt it's a suitable place.
And, IMO, we are not hard-coded here. We are defining compile time constants.

**Pull request merged**